### PR TITLE
use WebHID when available, fallback to BitBoxBridge

### PR DESCRIPTION
**Pull request by benma**
There is a hack regarding getDevicePath() - it now returns "WEBHID" if
there is WebHID in the browser. This is for backwards compatibility,
as clients usually call this function and feed the result to `new
BitBox02API()`.

To test WebHID: use Chrome
To test fallback to the bridge: use FireFox

### CODE REVIEW

File: src/bitbox02.js [hunk 1]
```diff
@@ -86,25 +93,93 @@ const setOutputDefaults = outputs =>  {
...
-        this.socket = null;
+        // connection is an object with three keys once the connection is established:
+        // onWrite(bytes): send bytes
+        // close():  close the connection
+        // valid(): bool - is the connection still alive?
+        this.connection = null;
+        this.onCloseCb = null;
+
+        if (navigator.hid) {
+            navigator.hid.addEventListener("disconnect", () => {
```
Comment (x1ddos) [inline src/bitbox02.js:110]: what if a user has 2 devices connected, one of which is bb02 and the other is something else. then they disconnect that something else: wouldn't it close the connection with bb02?

File: src/bitbox02.js [hunk 2]
```diff
@@ -86,25 +93,93 @@ const setOutputDefaults = outputs =>  {
...
+                if (this.onCloseCb) {
+                    this.onCloseCb();
+                }
+            };
             socket.onopen = function (event) {
-                resolve(socket);
+                resolve({
+                    onWrite: bytes => {
+                        if (socket.readyState != WebSocket.OPEN) {
+                            console.log("Error, trying to write to closed socket");
```
Comment (x1ddos) [inline src/bitbox02.js:132]: `console.error("attempted write to a closed socket")`? same in connectWebHID.

File: gowrapper/main.go [hunk 3]
```diff
@@ -137,12 +169,19 @@ func newJSDevice(onWrite func([]byte)) *js.Object {
...
 			},
-			close: func() {
+			close: func() {},
+		}
+	}
+	device := firmware.NewDevice(
+		nil,
+		nil,
+		&config{},
+		communication,
```
Comment (x1ddos) [inline gowrapper/main.go:183]: could've fit on one line

File: gowrapper/main.go [hunk 4]
```diff
@@ -110,13 +111,45 @@ func main() {
...
 	readChan := make(chan []byte)
-	device := firmware.NewDevice(
-		nil,
-		nil,
-		&config{},
-		&bb02Communication{
+	const bitboxCMD = 0x80 + 0x40 + 0x01
+
+	var communication firmware.Communication
+	if !useBitBoxBridge {
```
Comment (x1ddos) [inline gowrapper/main.go:137]: kind of weird to read. why the negation?

```go
if useBitBoxBridge {
   ...
} else {
  ...
}
```

File: gowrapper/main.go [hunk 5]
```diff
@@ -110,13 +111,45 @@ func main() {
...
+func (r readWriteCloser) Write(p []byte) (n int, err error) {
+	return r.write(p)
+}
+
+func (r readWriteCloser) Close() error { return nil }
+
+// useBitBoxBridge:
+// - if true, assume the connection is to the BitBoxBridge, which uses a slightly different communication protocol (no u2fhid frames).
+// - if false, assume the connection is via WebHID, so the regular u2fhid framing protocol is used.
+func newJSDevice(onWrite func([]byte), useBitBoxBridge bool) *js.Object {
```
Comment (x1ddos) [inline gowrapper/main.go:132]: couldn't find this function in https://github.com/digitalbitbox/bitbox02-api-js/blob/master/docs/methods.md.

**Review by x1ddos — APPROVED** [review]

File: gowrapper/main.go (hunk 5 above)
Comment (benma) [reply gowrapper/main.go:132]: It is not part of the public api - bitbox02.js contains the public API and it uses newJSDevice internally.

Comment (x1ddos) [reply gowrapper/main.go:132]: `js.Global.Set("bitbox02", ..., "New": newJSDevice, ...)` - is this not public? https://github.com/digitalbitbox/bitbox02-api-js/blob/cbcb920/gowrapper/main.go#L79

Comment (benma) [reply gowrapper/main.go:132]: No, only what's in `index.js` is public, which is mostly `BitBox02API`.

Comment (x1ddos) [reply gowrapper/main.go:132]: then I would split this into 2 functions: `newBridgedDevice` and `newWebHIDDevice`.

Comment (benma) [reply gowrapper/main.go]: done, ptal

File: src/bitbox02.js (hunk 2 above)
Comment (benma) [reply src/bitbox02.js]: done

File: src/bitbox02.js (hunk 1 above)
Comment (benma) [reply src/bitbox02.js:110]: It only fires if a device is disconnected that was previously connected with WebHID on the same page.

We only handle one connected device at the moment, so it should be okay.

The API here is really lacking, it is not obvious how to identify which device disconnected if you indeed had multiple. They should've done it like with the websocket, where the event is attached to the connection...

Comment (x1ddos) [reply src/bitbox02.js:110]: :+1:

**Comment by x1ddos** [comment]
LGTM

**Pull request merged**